### ISSUE TITLE
Singularity 管理员手册指南

### ISSUE THREAD

**Issue by snowsnail** [snowsnail/blog#1]
# Singularity 管理员手册指南

## 快速开始

本小节将介绍如何从零开始，快速构建一个singularity容器镜像。
下文将以构建TensorFlow应用框架为例，依次展开介绍。

**环境要求：**

- 安装singularity（参照第一章的内容）
- 工作目录`share`可用空间： >20GB
- 能够访问公网
- 操作系统：CentOS7

### 第一步 获取系统基础镜像

**说明： 由于`Tensorflow`框架官方在`ubuntu`系统上做过测试验证，所以容器镜像的基础操作系统采用`ubuntu`**

- 编写镜像定义文件 `/share/base.def`

    ```
    Bootstrap: library
    From: ubuntu:18.04
    
    %labels
        Author demo@example.org
        Version v0.0.1
    ```

- 执行构建镜像命令

    ```bash
    singularity build --notest /share/base.sif /share/base.def
    ```

### 第二步 配置Nvidia驱动

⚠️**注意：镜像内安装的Nvidia驱动一定要和超算环境里的nvidia驱动保持一致，否则容器将不能正常运行！**

nvidia驱动官网地址：https://www.nvidia.com/Download/Find.aspx?lang=en-us
选择相应GPU型号和操作系统版本等信息，下载对应版本的驱动，或复制下载链接替换下文`wget https://...` 部分的内容

- 编写镜像定义文件 `/share/base-nvidia.def`
    
    ```
    Bootstrap: localimage
    From: /share/base.sif
    
    %post
        mkdir /nvidia
        cd /nvidia/
        apt-get update
        apt-get install wget -y
        wget https://us.download.nvidia.cn/tesla/410.72/nvidia-diag-driver-local-repo-ubuntu1804-410.72_1.0-1_amd64.deb
        dpkg -i /nvidia/nvidia-diag-driver-local-repo-ubuntu1804-410.72_1.0-1_amd64.deb
        apt-get install gnupg2 -y
        apt-key add /var/nvidia-diag-driver-local-repo-410.72/7fa2af80.pub
        apt-get update
        DEBIAN_FRONTEND=noninteractive apt-get install keyboard-configuration -y
        apt-get install cuda-drivers -y
        rm -rf /nvidia
  ``` 

- 执行构建镜像命令

    ```bash
    singularity build --notest /share/base-nvidia.sif /share/base-nvidia.def
    ```

### 第三步 构建TensorFlow镜像

安装参考 [Tensorflow安装](https://www.tensorflow.org/install/)

- 创建可写的容器目录 `/share/tensorflow`

    ```
    $ sudo singularity build --sandbox /share/tensorflow /share/base-nvidia.sif
    ```

- 进入容器

    ```
    $ sudo singularity shell --writable /share/tensorflow
    ```

    输出如下：
    
    ```
    Singularity tensorflow:~>
    ```
    
- 安装`pip`工具

    ```bash
    Singularity tensorflow:~> apt-get update
    Singularity tensorflow:~> apt-get install python-pip python-dev -y
    ```

- 安装 `tensorflow`

    ```bash
    Singularity tensorflow:~> pip install --user --upgrade tensorflow
    Singularity tensorflow:~> pip install pandas
    ```

- 测试 `tensorflow`

    ```bash
    Singularity tensorflow:~> python -c "import tensorflow as tf; tf.enable_eager_execution(); print(tf.reduce_sum(tf.random_normal([1000, 1000])))"
    ```

- 创建共享目录 `/Share`

    ⚠️注意： 共享目录名称必须和超算环境的共享目录名称保持一致
    
    ```bash
    Singularity tensorflow:~> mkdir /Share
    ```
    
- 退出容器shell

     ```bash
     Singularity tensorflow:~> exit
     ```
    
- 容器格式转换

    ```bash
    $ sudo singularity build /share/tensorflow.sif /share/tensorflow
    ```

### 第四步 测试实例

- 编写tensorflow测试用例 `/share/tf-hello.py`

    ```python
    #!/usr/bin/env python
    # encoding: utf-8
    import tensorflow as tf
    
    tf.enable_eager_execution()
    tf.add(1, 2).numpy()
    
    hello = tf.constant('Hello, TensorFlow!')
    hello.numpy()
    ```

- 运行测试

    ```
    singularity exec /share/tensorflow.sif python /share/tf-hello.py
    ```

## 第一章 安装

本小节主要介绍CentOS下的安装操作步骤，其它操作系统的安装过程请参阅官网手册[Installation](https://www.sylabs.io/guides/3.0/user-guide/installation.html)

### 1.1 安装依赖环境

```bash
$ sudo yum update -y && \
    sudo yum groupinstall -y 'Development Tools' && \
    sudo yum install -y \
    openssl-devel \
    libuuid-devel \
    libseccomp-devel \
    wget \
    git \
    squashfs-tools
```

### 1.2 安装Go

- 下载`Go`安装包

    下载地址： [https://studygolang.com/dl](https://studygolang.com/dl)
    选择下载版本：go1.11.linux-amd64.tar.gz
    解压软件包： tar -C /usr/local -xzvf go1.11.linux-amd64.tar.gz
    删除软件包：rm go1.11.linux-amd64.tar.gz

- 配置`Go`环境

    ```bash
    $ echo 'export GOPATH=${HOME}/go' >> ~/.bashrc && \
        echo 'export PATH=/usr/local/go/bin:${PATH}:${GOPATH}/bin' >> ~/.bashrc && \
        source ~/.bashrc
    ```

- 安装`dep`

    ```bash
    $ go get -u github.com/golang/dep/cmd/dep
    ```

### 1.3 下载源代码

```bash
$ go get -d github.com/sylabs/singularity
```

### 1.4 编译安装

```bash
$ export VERSION=v3.0.1  && \
  cd $GOPATH/src/github.com/sylabs/singularity && \
  git fetch && \
  git checkout $VERSION 
```

```bash
$ ./mconfig --prefix=/opt/singularity && \
    make -C ./builddir && \
    sudo make -C ./builddir install
```

```bash
$ . /opt/singularity/etc/bash_completion.d/singularity
```

### 1.5 卸载

```
rm -rf /opt/singularity
```

### 2.1 构建容器的几种方法

Singulairty 支持多种方式制作容器。用户可以根据自己的喜好选择合适的方式。
Singularity 使用`build`命令来创建容器。 `build`命令接受一个目标地址作为输入并且生成容器作为输出。
目标地址定义了创建容器的方法，`build` 命令接受的目标地址包括：

- URI若以 *shub://* 开头， 以Singularity Hub (http://singularity-hub.org/)为来源创建容器
- URI若以 *docker://* 开头，以Docker Hub (http://singularity-hub.org/)为来源创建容器
- 本地已经存在的image文件的路径，从已有的image文件创建新的容器
- 本地的目录的路径, 从本地一个sandbox目录创建容器
- Singularity 清单文件路径, 从清单文件创建容器

另外`build`命令可以生成三种格式的image文件。 可以通过相应参数选项来指定输出的image的格式

- 用于生成环境下的压缩只读 *squashfs* 文件系统(默认)
- 用于交互式开发环境下可写 *ext3* file 文件系统(--writable 选项)
- 交互式开发环境下称为sandbox的可写目录结构 *(ch)root directory* (--sandbox 选项)

由于`build`命令接受本地的目标容器文件生成任意格式的新的image文件，所以我们可以将现有的容器文件转换成容易格式的容器文件。

### 2.2 从Container Library构建容器

```bash
$ sudo singularity build lolcow.simg library://sylabs-jms/testing/lolcow
```

- `lolcow.simg`: 构建的容器名称
- `library://sylabs-jms/testing/lolcow`: Container Library URI下载地址
- 默认情况下将创建一个压缩只读的SIF容器，如果想有可写权限可以使用`--sandbox`参数

### 2.3 从Docker Hub构建容器

```bash
$ sudo singularity build lolcow.sif docker://godlovedc/lolcow
```

### 2.4 创建一个可写的容器目录

```bash
$ sudo singularity build --sandbox lolcow/ library://sylabs-jms/testing/lolcow
```

```
$ sudo singularity shell --writable lolcow/
```
### 2.5 容器格式转换

- `development/` 目录转SIF格式文件 `production.sif`

    ```
    $ sudo singularity build production.sif development/
    ```

### 2.6 从singularity定义文件创建容器

- 编写定义文件 `lolcow.def`

```
Bootstrap: docker
From: ubuntu:16.04

%post
    apt-get -y update
    apt-get -y install fortune cowsay lolcat

%environment
    export LC_ALL=C
    export PATH=/usr/games:$PATH

%runscript
    fortune | cowsay | lolcat
```

- 执行创建命令

```bash
$ sudo singularity build lolcow.sif lolcow.def
```

### 2.7 `build` 命令更多选项用法

```bash
$ singularity build --help
```

## 第三章 容器定义文件

singularity容器定义文件简称`def文件`，解释了如何构建自定义的容器。它的内容描述包括了基本的OS或者从开始构建的基本容器，要安装的软件，运行时的环境变量设置，要从主机系统添加的文件以及容器元数据的详细信息。

### 3.1 概述

singularity定义文件分为两部分：

`Header`：标题描述了在容器内构建的核心操作系统。`Sections`：定义的其余部分由`Sections`：每个部分由一个`%`字符定义，后跟特定部分的名称。所有`Sections`都是可选的，def文件可能包含给定`Sections`的多个实例。

### 3.2 Header

`Header`应写在def文件的顶部。告诉Singularity构建容器的基本操作系统。它由几个关键字组成。

**Bootstrap** : 确定创建基本操作系统的引导代理程序
**From** : 确定基础操作系统的版本

```
Bootstrap: library
From: debian:7
```

下面是每个引导代理的关键字。详情查看[附录](https://www.sylabs.io/guides/3.0/user-guide/appendix.html)中的示例：

- library 
- docker 
- shub 
- localimage
- yum 
- debootstrap 
- arch 
- busybox 
- zypper 

### 3.3 Sections

实例文件：

```
Bootstrap: library
From: ubuntu:18.04

%setup
    touch /file1
    touch ${SINGULARITY_ROOTFS}/file2

%files
    /file1
    /file1 /opt

%environment
    export LISTEN_PORT=12345
    export LC_ALL=C

%post
    apt-get update && apt-get install -y netcat
    NOW=`date`
    echo "export NOW=\"${NOW}\"" >> $SINGULARITY_ENVIRONMENT

%runscript
    echo "Container was created $NOW"
    echo "Arguments received: $*"
    exec echo "$@"

%startscript
    nc -lp $LISTEN_PORT

%test
    grep -q NAME=\"Ubuntu\" /etc/os-release
    if [ $? -eq 0 ]; then
        echo "Container base is Ubuntu as expected."
    else
        echo "Container base is not Ubuntu."
    fi

%labels
    Author d@sylabs.io
    Version v0.0.1

%help
    This is a demo container used to illustrate a def file that uses all
    supported sections.
```

- `%setup`:安装基本操作系统后，在容器外部的主机系统上执行这部分命令 
- `%files`:将文件从主机系统复制到容器中
- `%environment`:设置环境变量
- `%post`: 安装基本操作系统后，在容器内部执行这部分命令
- `%runscript`: 容器运行时执行这部分命令
- `%startscript`: 容器实例启动时执行这部分命令
- `%test`: 在容器构建的最后执行这部分命令
- `%labels`: 容器的描述信息，通常是键值对形式
- `%help`: 记录容器的使用帮助文档

## 第四章 参考资料

- https://www.sylabs.io/guides/3.0/user-guide
- https://www.sylabs.io/guides/3.0/admin-guide
- https://hub.docker.com/
- https://cloud.sylabs.io/library
- https://github.com/sylabs/singularity
- https://github.com/sylabs/examples
- https://www.nvidia.com/download/driverResults.aspx/140708/en-us
- https://github.com/ufoym/deepo
- https://github.com/tensorflow/tensorflow
- https://www.tensorflow.org/install/pip?lang=python2
